pkg/geoipupdate/download/download.go

```diff
@@ -0,0 +1,149 @@
+// Package download provides a library for checking/downloading/updating mmdb files.
+package download
+
+import (
+	"context"
+	"crypto/md5"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"time"
+)
+
+const (
+	Extension = ".mmdb"
+
+	// zeroMD5 is the default value provided as an MD5 hash for a non-existent
+	// database.
+	zeroMD5 = "00000000000000000000000000000000"
+)
+
+type Downloader interface {
+	GetOutdatedEditions(ctx context.Context) ([]Metadata, error)
+	DownloadEdition(ctx context.Context, edition Metadata) error
+	MakeOutput() ([]byte, error)
+}
+
+// Download exposes methods needed to check for and perform update to a set of mmdb editions.
+type Download struct {
+	// accountID is the requester's account ID.
+	accountID int
+	// client is an http client responsible of fetching database updates.
+	client *http.Client
+	// databaseDir is the database download path.
+	databaseDir string
+	// editionIDs is the list of editions to be updated.
+	editionIDs []string
+	// licenseKey is the requester's license key.
+	licenseKey string
+	// oldEditionsHash holds the hashes of the previously downloaded mmdb editions.
+	oldEditionsHash map[string]string
+	// metadata holds the metadata pulled for each edition.
+	metadata []Metadata
+	// preserveFileTimes sets whether database modification times are preserved across downloads.
+	preserveFileTimes bool
+	// url points to maxmind servers.
+	url string
+
+	now     func() time.Time
+	verbose bool
+}
+
+// New initializes a new Downloader struct.
+func New(
+	accountID int,
+	licenseKey string,
+	url string,
+	proxy *url.URL,
+	databaseDir string,
+	preserveFileTimes bool,
+	editionIDs []string,
+	verbose bool,
```

Comment:
A thought I have with this is if we are changing the API significantly anyway, I wonder if we could clean up more of the mistakes from the existing API. IMO it was largely a mistake. And I think we are still seeing some leak through of that (understandably).

For example I wonder if we need the API to know about things like the database directory, "preserve file times", proxy, verbose, maybe more. Maybe the API could be simpler - give an edition ID and a date and take an io.Writer to write the database to. We'd still need to deal with writing files and such in geoipupdate, but perhaps that does not need to be public.

If we were designing the API from scratch, how would it ideally look? e.g. does the config need to be public? It seems like we are basically redoing it at this point anyway.

Probably needs more thought and discussion though.